File: frontend/src/api/sessionCheckpoints.ts
import { apiFetch } from "./client";

export type SessionCheckpointSummary = {
  id: string;
  name: string;
  schema_version: number;
  status: "active" | "archived";
  scene_id: string | null;
  active_combat_id: string | null;
  entry_count: number;
  entity_count: number;
  base_campaign_version: number;
  version: number;
  created_at: string;
  notes?: string | null;
};

export type SessionCheckpointDetail = SessionCheckpointSummary & {
  entries: Array<Record<string, unknown>>;
  snapshot?: Record<string, unknown>;
};

export type SessionCheckpointRestorePreview = {
  checkpoint_id: string;
  can_restore: boolean;
  conflicts: Array<Record<string, unknown>>;
  warnings: string[];
  change_summary: Record<string, unknown>;
};

export function listSessionCheckpoints(
  campaignId: string,
  signal?: AbortSignal,
): Promise<SessionCheckpointSummary[]> {
  return apiFetch<{ checkpoints: SessionCheckpointSummary[] }>(
    `/campaigns/${campaignId}/session-checkpoints`,
    { signal },
  ).then((result) => result.checkpoints);
}

export function getSessionCheckpoint(
  campaignId: string,
  checkpointId: string,
): Promise<SessionCheckpointDetail> {
  return apiFetch<SessionCheckpointDetail>(
    `/campaigns/${campaignId}/session-checkpoints/${checkpointId}`,
  );
}

export function createSessionCheckpoint(
  campaignId: string,
  input: {
    name: string;
    scene_id?: string | null;
    active_combat_id?: string | null;
    entries?: Array<Record<string, unknown>>;
    expected_campaign_version?: number;
    notes?: string | null;
  },
): Promise<SessionCheckpointSummary> {
  return apiFetch<SessionCheckpointSummary>(
    `/campaigns/${campaignId}/session-checkpoints`,
    { method: "POST", body: input },
  );
}

export function previewSessionCheckpointRestore(
  campaignId: string,
  checkpointId: string,
  input: { expected_campaign_version?: number; force?: boolean } = {},
): Promise<SessionCheckpointRestorePreview> {
  return apiFetch<SessionCheckpointRestorePreview>(
    `/campaigns/${campaignId}/session-checkpoints/${checkpointId}/restore-preview`,
    { method: "POST", body: input },
  );
}

export function restoreSessionCheckpoint(
  campaignId: string,
  checkpointId: string,
  input: {
    expected_campaign_version?: number;
    force?: boolean;
    idempotency_key: string;
  },
): Promise<{
  restored: boolean;
  checkpoint_id: string;
  campaign_id: string;
  restored_at: string;
  entries: Array<Record<string, unknown>>;
  change_summary: Record<string, unknown>;
}> {
  return apiFetch(
    `/campaigns/${campaignId}/session-checkpoints/${checkpointId}/restore`,
    { method: "POST", body: input },
  );
}

export function archiveSessionCheckpoint(
  campaignId: string,
  checkpointId: string,
  version: number,
): Promise<SessionCheckpointSummary> {
  return apiFetch(
    `/campaigns/${campaignId}/session-checkpoints/${checkpointId}/archive`,
    { method: "POST", body: { version } },
  );
}

File: backend/src/dnd_dm_assistant/api/routes/session_checkpoints.py
from __future__ import annotations

from typing import Annotated, Any

from fastapi import APIRouter, Depends, HTTPException, Query, Request, status

from dnd_dm_assistant.api.dependencies import get_session_checkpoint_service
from dnd_dm_assistant.api.session_checkpoint_schemas import (
    SessionCheckpointArchiveRequest,
    SessionCheckpointCreateRequest,
    SessionCheckpointRestoreRequest,
)
from dnd_dm_assistant.domain.campaign_state import StateNotFoundError, VersionConflict
from dnd_dm_assistant.infrastructure.database.session_checkpoint_service import (
    CheckpointConflictError,
    SessionCheckpointService,
)

router = APIRouter(
    prefix="/campaigns/{campaign_id}/session-checkpoints",
    tags=["session-checkpoints"],
)


def _request_id(request: Request) -> str:
    return str(getattr(request.state, "request_id", "unknown"))


def _call(fn: Any) -> Any:
    try:
        return fn()
    except StateNotFoundError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    except VersionConflict as exc:
        raise HTTPException(status_code=409, detail=str(exc)) from exc
    except CheckpointConflictError as exc:
        raise HTTPException(
            status_code=409,
            detail={"message": str(exc), "conflicts": exc.conflicts},
        ) from exc
    except ValueError as exc:
        raise HTTPException(status_code=422, detail=str(exc)) from exc


@router.get("")
def list_session_checkpoints(
    campaign_id: str,
    service: Annotated[SessionCheckpointService, Depends(get_session_checkpoint_service)],
    include_archived: bool = Query(default=False),
) -> dict[str, Any]:
    return {
        "checkpoints": _call(
            lambda: service.list_checkpoints(
                campaign_id, include_archived=include_archived
            )
        )
    }


@router.get("/current-state")
def get_current_session_state(
    campaign_id: str,
    service: Annotated[SessionCheckpointService, Depends(get_session_checkpoint_service)],
) -> dict[str, Any]:
    return _call(lambda: service.current_state(campaign_id))


@router.post("", status_code=status.HTTP_201_CREATED)
def create_session_checkpoint(
    campaign_id: str,
    body: SessionCheckpointCreateRequest,
    service: Annotated[SessionCheckpointService, Depends(get_session_checkpoint_service)],
) -> dict[str, Any]:
    return _call(
        lambda: service.create(
            campaign_id,
            name=body.name,
            scene_id=body.scene_id,
            active_combat_id=body.active_combat_id,
            entries=body.entries,
            expected_campaign_version=body.expected_campaign_version,
            notes=body.notes,
        )
    )


@router.get("/{checkpoint_id}")
def get_session_checkpoint(
    campaign_id: str,
    checkpoint_id: str,
    service: Annotated[SessionCheckpointService, Depends(get_session_checkpoint_service)],
) -> dict[str, Any]:
    return _call(lambda: service.get(campaign_id, checkpoint_id))


@router.post("/{checkpoint_id}/restore-preview")
def preview_session_checkpoint_restore(
    campaign_id: str,
    checkpoint_id: str,
    body: SessionCheckpointRestoreRequest,
    service: Annotated[SessionCheckpointService, Depends(get_session_checkpoint_service)],
) -> dict[str, Any]:
    return _call(
        lambda: service.preview_restore(
            campaign_id,
            checkpoint_id,
            expected_campaign_version=body.expected_campaign_version,
            force=body.force,
        )
    )


@router.post("/{checkpoint_id}/restore")
def restore_session_checkpoint(
    campaign_id: str,
    checkpoint_id: str,
    body: SessionCheckpointRestoreRequest,
    request: Request,
    service: Annotated[SessionCheckpointService, Depends(get_session_checkpoint_service)],
) -> dict[str, Any]:
    return _call(
        lambda: service.restore(
            campaign_id,
            checkpoint_id,
            expected_campaign_version=body.expected_campaign_version,
            force=body.force,
            idempotency_key=body.idempotency_key or _request_id(request),
        )
    )


@router.post("/{checkpoint_id}/archive")
def archive_session_checkpoint(
    campaign_id: str,
    checkpoint_id: str,
    body: SessionCheckpointArchiveRequest,
    service: Annotated[SessionCheckpointService, Depends(get_session_checkpoint_service)],
) -> dict[str, Any]:
    return _call(
        lambda: service.archive(
            campaign_id,
            checkpoint_id,
            expected_version=body.version,
        )
    )
